test/model/src/include/model/core.h

```diff
@@ -0,0 +1,406 @@
+/*-
+ * Public Domain 2014-present MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef MODEL_CORE_H
+#define MODEL_CORE_H
+
+#include <map>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace model {
+
+/*
+ * data_value --
+ *     The data value stored in the model used for keys and values. We use a generic class, rather
+ *     than a specific type such as std::string, to give us flexibility to change data types in the
+ *     future, e.g., if this becomes necessary to explore additional code paths.
+ */
+class data_value {
```

Comment:
Done—this cleaned up the code of this class quite a bit, which is great. I'm not sure if that's what you have intended, but I ended up using `std::optional` inside this class rather than replacing the class itself. The reason for that is that I'd like to not have the use of `std::string` hard-coded if we need to move away to use other types. The intention for this class is to be the model's equivalent of `WT_ITEM`.

Reply:
I wanted to make you aware that the functionality you were implementing was available gratis from the std lib.
Q When you say move away to use other types, am I correct to assume: support multiple types?

Reply:
Yes, my hope here would be to have a single type that can represent multiple types. For example, while I'm starting with string keys and values hard-coded, it would be useful to support integer keys and values, or even composite keys, such as consisting of several ints and a string.

I haven't thought about how I would implement this—if this is what we end up doing—but probably I would just serialize it into a string or a byte array in some way that would allow lexicographical comparison. But that's for later.

In the meantime, I just wanted to make sure that I wouldn't make it difficult to do this in the future. That's why I'm wondering if it would make sense to stick with a wrapper type.

Reply:
I suggest you have a look at `std::variant` and see if you can bake it in earlier rather than later.
https://en.cppreference.com/w/cpp/utility/variant

Reply:
Thanks for the suggestion—I wasn't aware of this class! I ended up rewriting `data_value` as `std::variant`'s subclass, so that I can extend it. I guess I those extensions are not really necessary... but anyways, hopefully this is now better.

I am a little unhappy that I couldn't find a better way to switch based on type than a bunch of `if` statements with `std::holds_alternative`. I tried `std::visit`, but I could not get it to work—from what I see online, I may need C++20 to be able to use it for my purposes.

Reply:
I'm wary of inheriting from the standard library, but I have *very* limited experience with `std::variant` and it seems like something that is done, so if you're happy with the result I hope that is a win.
For context: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2162r0.html